# Laboratorio 4

- [Mónica Salvatierra - 22249](https://github.com/alee2602)
- [Paula Barillas - 22764](https://github.com/paulabaal12)
- [Derek Arreaga - 22537](https://github.com/FabianKel)

#####  LINK DE REPOSITORIO
https://github.com/FabianKel/LAB4-IA

## Task 1 - Preguntas Teóricas

1. Explique con sus propias palabras, qué son los grafos computaciones y cuál es su importancia para el
cálculo de gradientes en aplicaciones como backpropagation

Gráfico acíclico dirigido, para representar la expresión matemática final y cada nodo representa subexpresiones intermedias.

* Ayuda a simplificar el proceso
* Ayuda a obtener más información de la naturaleza y estructura modular de los cálculos de gradientes

2. Detalle cuales son los componentes y pasos que conforman una red neuronal. Con esto en mente, ¿cómo
mejoraría el perceptrón que hizo en el laboratorio pasado?

**Componentes:**
* a

3. Investigue y explique cómo se selecciona el valor K usando el método de la silueta para el algoritmo de K-
Means. Explique las fórmulas (ecuaciones) que lo componen así como las asumpciones, si hay.

4. Investigue sobre Principal Component Analysis (PCA) y responda respecto a algoritmos como K-Means:
a. ¿Cómo podría ayudarle a mejorar la calidad de sus clusters cuando se usa K-Means?

## Task 2 - Ejercicios Prácticos

### Task 2.1 - K-Means y Segmentación de Clientes de Banco

### **Importar las librerías a utilizar**

In [9]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.discriminant_analysis import StandardScaler

### **Leer archivo de datos**

In [10]:
data = pd.read_csv("bank_transactions.csv")

### **Obtener las primeros 15 datos**

In [11]:
data.head(15)

TransactionID CustomerID CustomerDOB CustGender CustLocation  \
0             T1   C5841053     10/1/94          F   JAMSHEDPUR   
1             T2   C2142763      4/4/57          M      JHAJJAR   
2             T3   C4417068    26/11/96          F       MUMBAI   
3             T4   C5342380     14/9/73          F       MUMBAI   
4             T5   C9031234     24/3/88          F  NAVI MUMBAI   
5             T6   C1536588     8/10/72          F     ITANAGAR   
6             T7   C7126560     26/1/92          F       MUMBAI   
7             T8   C1220223     27/1/82          M       MUMBAI   
8             T9   C8536061     19/4/88          F      GURGAON   
9            T10   C6638934     22/6/84          M       MUMBAI   
10           T11   C5430833     22/7/82          M       MOHALI   
11           T12   C6939838      7/7/88          M       GUNTUR   
12           T13   C6339347     13/6/78          M    AHMEDABAD   
13           T14   C8327851      5/1/92          F        THANE   
14           T15   C7917151     24/3/78          M         PUNE   

    CustAccountBalance TransactionDate  TransactionTime  \
0             17819.05          2/8/16           143207   
1              2270.69          2/8/16           141858   
2             17874.44          2/8/16           142712   
3            866503.21          2/8/16           142714   
4              6714.43          2/8/16           181156   
5             53609.20          2/8/16           173940   
6               973.46          2/8/16           173806   
7             95075.54          2/8/16           170537   
8             14906.96          2/8/16           192825   
9              4279.22          2/8/16           192446   
10            48429.49          2/8/16           204133   
11            14613.46          2/8/16           205108   
12            32274.78          2/8/16           203834   
13            59950.44          1/8/16            84706   
14            10100.84          1/8/16            82253   

    TransactionAmount (INR)  
0                     25.00  
1                  27999.00  
2                    459.00  
3                   2060.00  
4                   1762.50  
5                    676.00  
6                    566.00  
7                    148.00  
8                    833.00  
9                    289.11  
10                   259.00  
11                   202.00  
12                 12300.00  
13                    50.00  
14                   338.00

In [12]:
print(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048567 entries, 0 to 1048566
Data columns (total 9 columns):
 #   Column                   Non-Null Count    Dtype  
---  ------                   --------------    -----  
 0   TransactionID            1048567 non-null  object 
 1   CustomerID               1048567 non-null  object 
 2   CustomerDOB              1045170 non-null  object 
 3   CustGender               1047467 non-null  object 
 4   CustLocation             1048416 non-null  object 
 5   CustAccountBalance       1046198 non-null  float64
 6   TransactionDate          1048567 non-null  object 
 7   TransactionTime          1048567 non-null  int64  
 8   TransactionAmount (INR)  1048567 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 72.0+ MB
None


### **Exploración de datos**

In [18]:
print(data.describe())
print(data.isnull().sum())

       TransactionID    CustomerID    CustGender  CustLocation  \
count   1.048567e+06  1.048567e+06  1.048567e+06  1.048567e+06   
mean    5.242830e+05  4.420843e+05  7.332216e-01  4.125395e+03   
std     3.026954e+05  2.553189e+05  4.493374e-01  2.385674e+03   
min     0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%     2.621415e+05  2.210755e+05  0.000000e+00  2.075000e+03   
50%     5.242830e+05  4.420960e+05  1.000000e+00  4.132000e+03   
75%     7.864245e+05  6.631645e+05  1.000000e+00  5.792000e+03   
max     1.048566e+06  8.842640e+05  3.000000e+00  9.355000e+03   

       CustAccountBalance  TransactionTime  TransactionAmount (INR)  
count        1.046198e+06     1.048567e+06             1.048567e+06  
mean         1.378708e-17     1.916886e-16            -2.306660e-17  
std          1.000000e+00     1.000000e+00             1.000000e+00  
min         -1.363327e-01    -3.064415e+00            -2.394521e-01  
25%         -1.307546e-01    -6.448761e-01            -

In [ ]:
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score

# Columnas categóricas
categorical_columns = ['CustomerID', 'CustGender', 'CustLocation', 'TransactionID']

# Encoding de variables categóricas
le = LabelEncoder()
for column in categorical_columns:
    data[column] = le.fit_transform(data[column])

# Balanceo del dataset (si es necesario)
# En este caso, no es necesario balancear el dataset ya que estamos realizando clustering

# Manejo de valores faltantes
data.fillna(data.mean(), inplace=True)

# Escalado de variables

scaler = StandardScaler()
numeric_features = data.select_dtypes(include=[np.number]).columns
data[numeric_features] = scaler.fit_transform(data[numeric_features])

# Selección de variables
X = data[numeric_features]

# Construcción del modelo

kmeans = KMeans(n_clusters=5, random_state=42)
kmeans.fit(X)

# Asignar los clusters a los datos
data['Cluster'] = kmeans.labels_

# Evaluación del modelo
# Utilizaremos la métrica de Silhouette Score para evaluar la calidad del clustering

score = silhouette_score(X, kmeans.labels_)
print(f'Silhouette Score: {score}')

       TransactionID    CustomerID    CustGender  CustLocation  \
count   1.048567e+06  1.048567e+06  1.048567e+06  1.048567e+06   
mean    5.242830e+05  4.420843e+05  7.332216e-01  4.125395e+03   
std     3.026954e+05  2.553189e+05  4.493374e-01  2.385674e+03   
min     0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%     2.621415e+05  2.210755e+05  0.000000e+00  2.075000e+03   
50%     5.242830e+05  4.420960e+05  1.000000e+00  4.132000e+03   
75%     7.864245e+05  6.631645e+05  1.000000e+00  5.792000e+03   
max     1.048566e+06  8.842640e+05  3.000000e+00  9.355000e+03   

       CustAccountBalance  TransactionTime  TransactionAmount (INR)  
count        1.046198e+06     1.048567e+06             1.048567e+06  
mean         1.378708e-17     1.916886e-16            -2.306660e-17  
std          1.000000e+00     1.000000e+00             1.000000e+00  
min         -1.363327e-01    -3.064415e+00            -2.394521e-01  
25%         -1.307546e-01    -6.448761e-01            -

TypeError: can only concatenate str (not "int") to str

#### Task 2.1.1 - K-Means

### Task 2.2 - Mixture Modelos y Segmentación de Clientes de Banco

#### Task 2.2.1 - Mixture Models